### PR TITLE
rework transaction `EffectHash`

### LINKED ISSUE #2808: Rework EffectHash
This is a tracking ticket for modifying the `EffectHash`:

- [x] Refactor TransactionBody such that its component pieces can be incrementally hashed (e.g. by a hardware wallet): #2784
- [x] Combine those fields into the full transaction effect hash e.g. `H(fixed size hashes of per tx data packets || num actions || fixed size hashes per action)`
- [x] Use the protos to encode the effect hash for individual actions: #2000, ensuring for each action that where needed the current data modeling separates the effecting data (e.g. SpendBody vs Spend), and that only that field is included in the effect hash - https://github.com/penumbra-zone/penumbra/pull/2931

### PR DESCRIPTION
**Pull request by redshiftzero**
Closes #2808
Adds a section of documentation on transaction signing which describe the changes made in this PR

### CODE REVIEW

**Comment by redshiftzero** [comment]
CI is failing here due to unrelated proto changes, but this is ready for review

File: docs/protocol/src/crypto/transaction_signing.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+Each proto has a unique string associated with it which we call its *Type URL*,
+which is included in the inputs to BLAKE2b-512.
+Type URLs are variable length, so a fixed-length field (8 bytes) is first included
+in the hash to denote the length of the Type URL field.
+
+Summarizing the above, the effect hash _for each action_ is computed as:
+
+```
+effect_hash = BLAKE2b-512(len(type_url) || type_url || proto_encode(proto))
+```
```
Comment (cronokirby) [inline docs/protocol/src/crypto/transaction_signing.md:44]: Might want to specify how the length is encoded

**Review by cronokirby — APPROVED** [review]
Looks all fine and dandy

File: crates/core/transaction/src/effect_hash.rs [hunk 2]
```diff
@@ -54,38 +56,42 @@ impl TransactionBody {
...
     pub fn effect_hash(&self) -> EffectHash {
-        let mut state = blake2b_simd::Params::default()
-            .personal(b"PAH:tx_body")
-            .to_state();
+        let personalization = TransactionBody::TYPE_URL;
+
+        let mut state = blake2b_simd::State::new();
+        let length = personalization.len() as u64;
+        state.update(&length.to_le_bytes());
+        state.update(personalization.as_bytes());
```
[the remark this answers is not in the document]
Comment (hdevalence) [reply crates/core/transaction/src/effect_hash.rs:64]: Aesthetically I think it'd be nice if this used a u16 length for compactness but I'm not sure that that's actually a good idea. Probably best to not have to worry about the length.

**Pull request merged**